enverus_direct_access/__init__.py

```diff
@@ -523,3 +442,154 @@ def query(self, dataset, **options):
+            raise DAAuthException(
+                "API_KEY, CLIENT_ID and CLIENT_SECRET are required to generate an access token"
+            )
+        self.session.headers["Authorization"] = "Basic {}".format(
+            base64.b64encode(
+                ":".join([self.client_id, self.client_secret]).encode()
+            ).decode()
+        )
+        self.session.headers["Content-Type"] = "application/x-www-form-urlencoded"
+
+        payload = {"grant_type": "client_credentials"}
+        response = self.session.post(url, params=payload)
+        self.logger.debug("Token response: " + json.dumps(response.json(), indent=2))
+        self.access_token = response.json()["access_token"]
+        self.session.headers["Authorization"] = "bearer {}".format(self.access_token)
+        return response.json()
+
+
+class DeveloperAPIv3(BaseAPI):
+    """Client for Enverus' Developer API Version 3"""
+
+    def __init__(
+            self,
+            api_key,
+            retries=5,
+            backoff_factor=1,
+            links=None,
+            access_token=None,
+            **kwargs
+    ):
+        """
+        Enverus' Developer API Version 3 client
+
+        API documentation and credentials can be found at: https://app.enverus.com/direct/#/api/explorer/v3/gettingStarted
+
+        :param api_key: api key credential.
+        :type api_key: str
+        :param retries: the number of attempts when retrying failed requests with status codes of 500, 502, 503 or 504
+        :type retries: int
+        :param backoff_factor: the factor to use when exponentially backing off prior to retrying a failed request
+        :type backoff_factor: int
+        :param links: a dictionary of prev and next links as provided by the python-requests Session.
+        See https://requests.readthedocs.io/en/master/user/advanced/#link-headers
+        :type dict
+        :param access_token: an optional, pregenerated access token. If provided, the class instance will not
+        automatically try to request a new access token.
+        :type: access_token: str
+        :param kwargs:
+        """
+        super(DeveloperAPIv3, self).__init__(api_key, retries, backoff_factor, **kwargs)
+        self.links = links
+        self.access_token = access_token
+        self.url = self.url + "/v3/direct-access"
+        self.session.hooks["response"].append(self._check_response)
+
+        if self.access_token:
+            self.session.headers["Authorization"] = "bearer {}".format(
+                self.access_token
+            )
+        else:
+            self.access_token = self.get_access_token()["token"]
+
+    def get_access_token(self):
+        """
+        Get an access token from /tokens endpoint. Automatically sets the Authorization header on the class instance's
+        session. Raises DAAuthException on error
+
+        :return: token response as dict
+        """
+        url = self.url + "/tokens"
```

Comment:
Do not use "+" for string concatenation, use os.path.join for url parts concatenation